docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,76 @@
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
+
+## Implementation Plan
+
+```c++
+enum ConnectionManagerStates {
+  InternetProbe,
+  Firewall,
+  ServerProbe,
+  CheckSubscription,
+  Idle,
+}
+```
+### Upon VPN Activation (`StateOff` to `StateConnecting`)
+Perform a single cycle going through all the Connection Manager probes.
+- If a probe fails:
+  - Set Controller to `StateOff`
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Activate VPN (Controller `StateOn`)
+
+### VPN is already active (`StateOn`)
+Perform periodic checks cycling through all the Connection Manager probes.
+- If a probe fails:
+  - Controller continues to stay in `StateOn`. We will __never__ toggle the VPN off without the user explicitly doing so themselves
+  - We will __not__ manipulate the Connection Health stability based on Connection Manager probes
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Do nothing. VPN stays on and stable
+
+> Note: The second scenario where a check fails when the VPN is already active still needs some more thinking. There will likely be a follow up design doc on how we will approach that later on.
+
+### How to recover after a failed probe
+- If VPN is not in `StateOn` yet:
+  - Connection Manager will emit signal about the failed probe to notify the frontend
+- If VPN is in `StateOn`:  
+  - Wait until Connection Health stability goes back to Stable or Unstable
+  - Restart the sequence of ConnectionManager checks
+
+![Diagram to explain how Controller, Connection Manager and Connection Health states interact](./CM_diagram.png)
+
+## Metrics and Measurements
+
+This change also means that prior to the activation of the VPN, there are multiple checks and probes that need to succeed; this will take some time (hopefully no more than a couple of seconds).
+
+To confidently determine that these checks do not delay the VPN activation by an unacceptable amount TBD, we need to measure the duration of the _time elapsed from toggling on the VPN to activation_. We are interested in the VPN activation duration because it delays the VPN being toggled on. This also allows us to keep track of the duration of each operation which allows us to continually monitor the system health and identify any slow downs and changes.
```

Comment:
Will this confirmation that it doesn't introduce an unacceptable delay be done in testing? By QA? Based on early metrics we get back from users?

Reply:
Likely both

Reply:
Can that be explicitly mentioned in the doc?